torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+
+            # Don't free the full params for the outer-most (root) instance,
+            # since those params will be needed immediately after for the
+            # backward pass.
+            self.reshard_after_forward = False
+
+            # Due to the use of streams, we need to make sure the previous
+            # ``optim.step()`` is done before we all-gather parameters.
+            self._wait_for_previous_optim_step()
+
+    @torch.no_grad()
+    def _init_param_attributes(self, p: Parameter) -> None:
+        """
+        We manage several attributes on each Parameter instance. The first two
+        are set by :func:`_shard_parameters_`:
+            ``_is_sharded``: ``True`` if the Parameter is sharded or ``False``
+                if the Parameter is intentionally not sharded (in which case we
+                will all-reduce grads for this param).
+            ``_orig_size``: the size of the original Parameter (before sharding)
+        The remaining attributes are set here:
+            ``_local_shard``: a single shard of the parameters.
+            ``_full_param_padded``: the full weight (padded to be evenly
+                divisible by ``world_size``), used for computation in the
+                forward and backward pass. This will be resized in place and
+                only materialized (via all-gather) as needed.
+        """
+        assert hasattr(p, "_is_sharded") and hasattr(p, "_orig_size")
+        if hasattr(p, "_local_shard"):
+            return
+
+        # A single shard of the parameters.
+        p._local_shard = p.data  # type: ignore[attr-defined]
+
+        # We also maintain a full-sized parameter of type self.compute_dtype.
+        # We resize the storage to size 0 at init (here) and only materialize
+        # as needed. The storage may contain padding elements so that it is
+        # evenly divisible by world_size, although these padding elements will
+        # be removed before the relevant computation.
+        if p._is_sharded:  # type: ignore[attr-defined]
+            p._full_param_padded = torch.zeros(  # type: ignore[attr-defined]
+                p.numel() * self.world_size, device=self.compute_device, dtype=self.compute_dtype
+            )
+            _free_storage(p._full_param_padded)  # type: ignore[attr-defined]
+
+    def _set_is_root(self) -> None:
+        """If ``True``, implies that no other :class:`FullyShardedDataParallel`
+        instance wraps this one. Called once by :func:`_lazy_init`.
+        """
+        if self._is_root is not None:
+            return
+        # No FSDP instance wraps this, else _is_root would be set to False.
+        self._is_root = True
+        # If final backward callback is never been queued, state should be IDLE.
+        # If final backward callback is queued, the callback should be finished
+        # and the state was reset to be IDLE.
+        # This should be asserted at the beginning of forward pass in the root instance only.
```

Comment:
line wrap length is different in different part of this file. Could you please run some formatting tool to format the code?